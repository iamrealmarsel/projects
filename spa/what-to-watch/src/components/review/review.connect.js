import {connect} from 'react-redux';
import {postComment} from 'store/actions/async';


const mapStateToProps = ({load}) => ({
  movies: load.movies,
});

const mapDispatchToProps = {
  postCommentAction: postComment,
};


export default connect(mapStateToProps, mapDispatchToProps);
