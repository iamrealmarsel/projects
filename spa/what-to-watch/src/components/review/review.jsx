import React from 'react';
import PropTypes from 'prop-types';
import connect from 'components/review/review.connect';
import RatingStars from 'components/rating-stars/rating-stars';
import Header from 'components/header/header';
import {isSmallAmountOfText, isTextLimitExceeded} from 'utils';
import {moviesPropTypes} from 'store/prop-types';
import {getMovieByID} from 'store/selector';
import {SHAKE_ANIMATION_TIMEOUT} from 'const';


const formDataState = {
  ratingStarsChecks: [false, false, true, false, false],
  text: ``,
};

const formDisabledState = {
  postButton: true,
  textArea: false,
};

const Review = (props) => {
  const {movies, postCommentAction} = props;
  const id = Number(props.match.params.id);
  const currentMovie = getMovieByID(movies, id);
  const {title, background, poster} = currentMovie;

  const [formData, setData] = React.useState(formDataState);
  const [formDisabled, setIsDisabled] = React.useState(formDisabledState);
  const [errorShake, setIsError] = React.useState(false);

  const handleStarClick = (checked, index) => {
    const newRatingStarsCheked = Array(5).fill(false);
    newRatingStarsCheked[index] = checked;
    setData((prevState) => Object.assign({}, prevState, {ratingStarsChecks: newRatingStarsCheked}));
  };

  const handleTextChange = (text) => {
    if (isTextLimitExceeded(text)) {
      return;
    }
    setData((prevState) => Object.assign({}, prevState, {text}));
    setIsDisabled((prevState) => Object.assign({}, prevState, {postButton: isSmallAmountOfText(text)}));
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    const rating = formData.ratingStarsChecks.indexOf(true) + 1;
    const text = formData.text;

    setIsDisabled((prevState) => Object.assign({}, prevState, {
      postButton: true,
      textArea: true,
    }));

    postCommentAction(rating, text, id)
      .then(() => {
        setIsDisabled((prevState) => Object.assign({}, prevState, {
          postButton: false,
          textArea: false,
        }));
      })
      .catch(() => {
        setIsDisabled((prevState) => Object.assign({}, prevState, {
          postButton: false,
          textArea: false,
        }));
        reportError();
      });
  };

  const reportError = () => {
    setIsError(true);
    setTimeout(() => setIsError(false), SHAKE_ANIMATION_TIMEOUT);
  };

  return (
    <section className="movie-card movie-card--full">
      <div className="movie-card__header">
        <div className="movie-card__bg">
          <img src={background} alt={title} />
        </div>

        <h1 className="visually-hidden">WTW</h1>

        <Header id={id} movieTitle={title} />

        <div className="movie-card__poster movie-card__poster--small">
          <img src={poster} alt={title} width="218" height="327" />
        </div>
      </div>

      <div className="add-review">
        <form action="" className={`add-review__form ${errorShake ? `shake` : ``} `} onSubmit={handleSubmit} >
          <div className="rating">
            <div className="rating__stars">
              {formData.ratingStarsChecks.map((starCheck, index) => (
                <RatingStars
                  key={index}
                  index={index}
                  starCheck={starCheck}
                  onStarClick={(checked) => handleStarClick(checked, index)} />
              ))}
            </div>
          </div>

          <div className="add-review__text">
            <textarea
              className="add-review__textarea"
              name="review-text" id="review-text"
              placeholder="Review text"
              onChange={(event) => handleTextChange(event.target.value)}
              disabled={formDisabled.textArea}
              value={formData.text}
            >
            </textarea>
            <div className="add-review__submit">
              <button className="add-review__btn" type="submit" disabled={formDisabled.postButton} >Post</button>
            </div>

          </div>
        </form>
      </div>

    </section>
  );
};

Review.propTypes = {
  movies: moviesPropTypes,
  postCommentAction: PropTypes.func.isRequired,
  match: PropTypes.object.isRequired
};


export {Review};
export default connect(Review);
