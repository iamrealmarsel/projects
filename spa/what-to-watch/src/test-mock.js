import {createStore} from 'redux';
import rootReducer from 'store/root-reducer';
import {loadMoviePromo} from 'store/actions/load';


export const noop = () => {};
export const VISIBLE_MOVIES_COUNT = 16;
export const ID = 1;
export const RATING_STARS = [false, false, true, false, false];
export const TAB = `details`;
export const PREVIEW = `img/aviator.jpg`;
export const PREVIEW_VIDEO = `https://upload.wikimedia.org/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`;
export const TITLE = `Bohemian Rhapsody`;
export const CURRENT_GENRE = `Comedy`;
export const ALL_GENRES = `All genres`;
export const GENRES = [`Comedy`, `Crime`, `Documentary`, `Drama`, `Horror`, `Kids & Family`];
export const MATCH = {params: {id: 1}};
export const PARAMS = {id: 1};

export const MOVIES = [
  {
    description: `A young man who was sentenced to seven years`,
    rating: 3.6,
    director: `Nicolas Winding Refn`,
    genre: `Action`,
    starring: [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
    id: 1,
    title: `Bronson`,
    poster: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    preview: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    background: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    backgroundColor: `#73B39A`,
    previewVideo: `https://upload.wikimedia.org/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    video: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    votes: 109661,
    runtime: 92,
    releaseYear: 2008,
    myList: false,
    comments: false,
  },
  {
    description: `A young man who was sentenced to seven years`,
    rating: 3.6,
    director: `Nicolas Winding Refn`,
    genre: `Action`,
    starring: [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
    id: 2,
    title: `Bronson`,
    poster: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    preview: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    background: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    backgroundColor: `#73B39A`,
    previewVideo: `https://upload.wikimedia.org/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    video: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    votes: 109661,
    runtime: 92,
    releaseYear: 2008,
    myList: false,
    comments: false,
  },
];

export const MOVIE = {
  description: `A young man who was sentenced to seven years`,
  rating: 3.6,
  director: `Nicolas Winding Refn`,
  starring: [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
  genre: `Action`,
  id: 3,
  title: `Bronson`,
  poster: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
  preview: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
  background: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
  backgroundColor: `#73B39A`,
  previewVideo: `https://upload.wikimedia.org/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  video: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  votes: 109661,
  runtime: 92,
  releaseYear: 2008,
  myList: false,
  comments: false,
};

export const REVIEWS = [
  {
    id: 111,
    rating: 3.3,
    date: `2020-10-09T13:38:44.769Z`,
    author: `Christina`,
    message: `This film is an experience and i has already seen it 4 times`,
  },
  {
    id: 222,
    rating: 3.3,
    date: `2020-10-09T13:38:44.769Z`,
    author: `Christina`,
    message: `This film is an experience and i has already seen it 4 times`,
  }
];

const store = createStore(rootReducer);
store.dispatch(loadMoviePromo(MOVIE));
export {store};
