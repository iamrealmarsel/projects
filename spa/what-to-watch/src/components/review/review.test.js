import React from "react";
import TestRenderer from "react-test-renderer";
import {BrowserRouter} from 'react-router-dom';
import {Provider} from 'react-redux';
import {Review} from "./review";
import * as mock from "test-mock";


test(`render Review`, () => {
  const tree = TestRenderer
    .create(
        <Provider store={mock.store}>
          <BrowserRouter>
            <Review
              movies={mock.MOVIES}
              postCommentAction={mock.noop}
              match={mock.MATCH}
            />
          </BrowserRouter>
        </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
