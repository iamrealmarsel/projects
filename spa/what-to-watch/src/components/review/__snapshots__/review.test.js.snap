// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Review 1`] = `
<section
  className="movie-card movie-card--full"
>
  <div
    className="movie-card__header"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="Bronson"
        src="https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header user-page__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <nav
        className="breadcrumbs"
      >
        <ul
          className="breadcrumbs__list"
        >
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
              href="/films/1"
              onClick={[Function]}
            >
              Bronson
            </a>
          </li>
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
            >
              Add review
            </a>
          </li>
        </ul>
      </nav>
      <div
        className="user-block"
      >
        <a
          className="user-block__link"
          href="/login"
          onClick={[Function]}
        >
          Sign in
        </a>
      </div>
    </header>
    <div
      className="movie-card__poster movie-card__poster--small"
    >
      <img
        alt="Bronson"
        height="327"
        src="https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg"
        width="218"
      />
    </div>
  </div>
  <div
    className="add-review"
  >
    <form
      action=""
      className="add-review__form  "
      onSubmit={[Function]}
    >
      <div
        className="rating"
      >
        <div
          className="rating__stars"
        >
          <input
            checked={false}
            className="rating__input"
            id="star-0"
            name="rating"
            onChange={[Function]}
            type="radio"
            value={0}
          />
          <label
            className="rating__label"
            htmlFor="star-0"
          >
            Rating 
            0
          </label>
          <input
            checked={false}
            className="rating__input"
            id="star-1"
            name="rating"
            onChange={[Function]}
            type="radio"
            value={1}
          />
          <label
            className="rating__label"
            htmlFor="star-1"
          >
            Rating 
            1
          </label>
          <input
            checked={true}
            className="rating__input"
            id="star-2"
            name="rating"
            onChange={[Function]}
            type="radio"
            value={2}
          />
          <label
            className="rating__label"
            htmlFor="star-2"
          >
            Rating 
            2
          </label>
          <input
            checked={false}
            className="rating__input"
            id="star-3"
            name="rating"
            onChange={[Function]}
            type="radio"
            value={3}
          />
          <label
            className="rating__label"
            htmlFor="star-3"
          >
            Rating 
            3
          </label>
          <input
            checked={false}
            className="rating__input"
            id="star-4"
            name="rating"
            onChange={[Function]}
            type="radio"
            value={4}
          />
          <label
            className="rating__label"
            htmlFor="star-4"
          >
            Rating 
            4
          </label>
        </div>
      </div>
      <div
        className="add-review__text"
      >
        <textarea
          className="add-review__textarea"
          disabled={false}
          id="review-text"
          name="review-text"
          onChange={[Function]}
          placeholder="Review text"
          value=""
        />
        <div
          className="add-review__submit"
        >
          <button
            className="add-review__btn"
            disabled={true}
            type="submit"
          >
            Post
          </button>
        </div>
      </div>
    </form>
  </div>
</section>
`;
